### PR TITLE
QA-232: suggest cherry-picks to the latest three release branches

### PR DESCRIPTION
**Pull request by tranchitella**

### CODE REVIEW

**Comment by tranchitella** [comment]
@lluiscampos any suggestion on how to get rid of the `Host key verification failed` (see the failed job in the CI pipeline)?

**Comment by tranchitella** [comment]
If we want to run unit tests, I need the goveralls setup. Before proceeding, though, I'm waiting for your feedbacks @lluiscampos and @oleorhagen

File: menderqa_pipeline_helpers.go [hunk 1]
```diff
@@ -9,8 +9,21 @@ import (
...
+	if err != nil {
+		err = fmt.Errorf("getLatestIntegrationRelease: Error: %v (%s)", err, version)
+	}
+	versionStr := strings.TrimSpace(string(version))
+	return strings.SplitN(versionStr, "\n", -1), err
+}
+
 func getServiceRevisionFromIntegration(repo, baseBranch string, conf *config) (string, error) {
-	c := exec.Command("release_tool.py", "--version-of", repo, "--in-integration-version", baseBranch)
+	c := exec.Command("python3", "release_tool.py", "--version-of", repo, "--in-integration-version", baseBranch)
```
Comment (oleorhagen) [inline menderqa_pipeline_helpers.go:26]: Question:
Why was this needed?

**Review by oleorhagen — APPROVED** [review]
I think it looks good.

Prior to reviewing this I was thinking along the lines, that maybe we should have the ability to modify the predicate it matches on for notifying about cherries needed (like a function which takes the commit message (string), and returns true of false, and is then mapped onto the commit messages in order). However, it's not necessary. 

Also, prior to reviewing I wasn't sure if `Changelogs` are the best measure here. But you guys are probably right (I was thinking along the lines of having something like `fix` or `fixes` in the commit message would trigger this.).

File: cherrypicks.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+	gitcmd = exec.Command("git", "fetch", "github", "pull/"+prNumber+"/head:"+prBranchName)
+	gitcmd.Dir = tmpdir
+	out, err = gitcmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("%v returned error: %s: %s", gitcmd.Args, out, err.Error())
+	}
+
+	// count the number of non-none changelog entries
+	baseSHA := *pr.PullRequest.Base.SHA
+	countCmd := exec.Command("sh", "-c", "git log "+baseSHA+"...pr_"+prNumber+" | grep -i \"Changelog: \" | grep -v -i \": none\" | wc -l")
```
Comment (kacf) [inline cherrypicks.go:79]: Suggest being a bit more explicit, since it's not uncommon to include the word "changelog" in various other contexts for example when doing releases, and to have the word "none" in an actual changelog message:

```
grep -i \"^Changelog:\" | grep -v -i \"^Changelog: *none\"
```

Might be useful to have a test for these sentences too. Examples would be:

* ```
  Fix changelog: This was actually reverted.

  Cancel-changelog: abcdef0123456789
   ```
* ```
  Check for nil.

  Changelog: Fix API returning none.
```

File: deb-requirements.txt [hunk 3]
```diff
@@ -0,0 +1 @@
+python3-yaml
```
[the remark this answers is not in the document]
Comment (kacf) [reply deb-requirements.txt:1]: Not if the dependency on this component is introduced in this commit.

File: menderqa_pipeline_helpers.go [hunk 4]
```diff
@@ -9,8 +9,21 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
+func getLatestIntegrationRelease(number int, conf *config) ([]string, error) {
+	cmd := fmt.Sprintf("git for-each-ref --sort=-creatordate --format='%%(refname:short)' 'refs/tags' "+
+		"| sed -E '/(^[0-9]+\\.[0-9]+)\\.[0-9]+$/!d;s//\\1.x/' | uniq | head -n %d | sort -n -r", number)
```
Comment (kacf) [inline menderqa_pipeline_helpers.go:14]: Suggestion: Instead of `sort -n`, use `sort -V`, which is specifically for sorting versions.

File: deb-requirements.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (tranchitella) [reply deb-requirements.txt:1]: I need it to run unit tests, why do I need to separate it in a different commit?

File: entfork_sync.go [hunk 5]
```diff
@@ -14,8 +14,9 @@ import (
 	"text/template"
 
 	"github.com/google/go-github/v28/github"
-
 	"github.com/sirupsen/logrus"
+
+	tgithub "github.com/mendersoftware/integration-test-runner/client/github"
```
[the remark this answers is not in the document]
Comment (tranchitella) [reply entfork_sync.go:19]: because "github" is the import of the GitHub library, I need to use an alias because it is clashing with my github client, which I introduced to be able to mock it in unit tests.

File: main_test.go [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+	cmd := exec.Command("sh", "-c", "! [ -f $HOME/.ssh/known_hosts ] && "+
+		"mkdir -p $HOME/.ssh && ssh-keyscan github.com | "+
+		"tee -a $HOME/.ssh/known_hosts && chmod 600 $HOME/.ssh/known_hosts || true")
+	err := cmd.Run()
+	if err != nil {
+		panic(err)
+	}
+	// import the SSH private key
+	cmd = exec.Command("sh", "-c", "[ -n \"$SSH_PRIVATE_KEY\" ] && ! [ -f $HOME/.ssh/id_rsa ] && "+
+		"echo \"$SSH_PRIVATE_KEY\" | tr -d '\r' | tee $HOME/.ssh/id_rsa && chmod 600 $HOME/.ssh/id_rsa || true")
```
Comment (kacf) [inline main_test.go:18]: We should not use `tee` for a private key, suggest using straight redirection so it doesn't show up in the log.

File: menderqa_pipeline_helpers.go (hunk 1 above)
Comment (tranchitella) [reply menderqa_pipeline_helpers.go:26]: Because it doesn't work if you don't have "." in your PATH (like on my Mac OS).
Unit tests will fail because release_tool.py is not executable directly. It's either "./release_tool.py" or "python3 release_tool.py".

File: deb-requirements.txt (hunk 3 above)
[the remark this answers is not in the document]
Comment (oleorhagen) [reply deb-requirements.txt:1]: I see, I figured it was missing from old.

File: entfork_sync.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (oleorhagen) [reply entfork_sync.go:19]: Yes, but `t` for `temp`? :thinking:

[the remark this answers is not in the document]
Comment (tranchitella) [reply entfork_sync.go:19]: t for `test-runner`, that was my intention :)

[the remark this answers is not in the document]
Comment (tranchitella) [reply entfork_sync.go:19]: let's rename it to client_github

File: README.md [hunk 7]
```diff
@@ -21,35 +21,29 @@ For all repositories in the organization, a pr_XXX branch will be created in
 GitLab for every pull/XXX PR from GitHub.
```
Comment (lluiscampos) [inline README.md:22]: Thanks for updating README

File: cherrypicks.go [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+	countCmd.Dir = tmpdir
+	out, err = countCmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("%v returned error: %s: %s", gitcmd.Args, out, err.Error())
+	}
+
+	changelogs, _ := strconv.Atoi(strings.TrimSpace(string(out)))
```
[the remark this answers is not in the document]
Comment (tranchitella) [reply cherrypicks.go:86]: TBH, I like strconv.Atoi. Can you elaborate more on the advantages of binary/encoding?

File: cherrypicks.go (hunk 2 above)
Comment (tranchitella) [reply cherrypicks.go]: Nice, thanks! Your grep machinery is way better than mine, I applied it.

File: menderqa_pipeline_helpers.go [hunk 9]
```diff
@@ -9,8 +9,21 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
+func getLatestIntegrationRelease(number int, conf *config) ([]string, error) {
```
Comment (lluiscampos) [inline menderqa_pipeline_helpers.go:12]: This function does not belong here: `menderqa_pipeline_helpers.go` are meant to be used only from `menderqa_pipeline.go`. Maybe you want to put it in `util.go`, or directly in `cherrypicks.go` (it is not used anywhere else).

File: main.go [hunk 10]
```diff
@@ -264,14 +251,19 @@ func processGitHubWebhook(ctx *gin.Context, payload []byte, githubClient *github
...
 		// Keep the OS and Enterprise repos in sync
-		if err = syncIfOSHasEnterpriseRepo(log, conf, pr); err != nil {
+		if err := syncIfOSHasEnterpriseRepo(log, conf, pr); err != nil {
 			log.Errorf("Failed to sync the OS and Enterprise repos: %s", err.Error())
 		}
+
+		// If the pr was merged, suggest cherry-picks
+		if err := suggestCherryPicks(log, pr, githubClient, conf); err != nil {
+			log.Errorf("Failed to suggest cherry picks for the pr %v. Error: %v", pr, err)
+		}
```
Comment (lluiscampos) [inline main.go:266]: This has to be done above line 240, or else won't be processed for PRs from contributors.

File: main_test.go [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+		"mkdir -p $HOME/.ssh && ssh-keyscan github.com | "+
+		"tee -a $HOME/.ssh/known_hosts && chmod 600 $HOME/.ssh/known_hosts || true")
+	err := cmd.Run()
+	if err != nil {
+		panic(err)
+	}
+	// import the SSH private key
+	cmd = exec.Command("sh", "-c", "[ -n \"$SSH_PRIVATE_KEY\" ] && ! [ -f $HOME/.ssh/id_rsa ] && "+
+		"echo \"$SSH_PRIVATE_KEY\" | tr -d '\r' >> $HOME/.ssh/id_rsa && chmod 600 $HOME/.ssh/id_rsa || true")
+	err = cmd.Run()
```
Comment (lluiscampos) [inline main_test.go:19]: I don't understand what is the intention here. We are not going to set an SSH_PRIVATE_KEY in CI variables, so is only meant for local use? Can the tests run without the SSH key at all?

**Review by lluiscampos — CHANGES_REQUESTED** [review]
Having already @kacf and @oleorhagen eyes on the code, I focused more on high level things.

See my comments below; the only one "requesting for changes" is the one in `main.go`, as I believe we need to move this check earlier in the process.

Thank you very much for introducing unit tests. This is going to help everyone working on the tool :heart:

File: cherrypicks.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (oleorhagen) [reply cherrypicks.go:86]: You just don't have to go via strings, but as mentioned: `feel free to choose`. And it seems you have :)

File: main_test.go (hunk 11 above)
Comment (tranchitella) [reply main_test.go:19]: Without the SSH key, the unit tests cannot clone the repos, thus unit tests won't run in CI. Am I wrong?

File: menderqa_pipeline_helpers.go (hunk 9 above)
Comment (tranchitella) [reply menderqa_pipeline_helpers.go]: moved

File: main.go (hunk 10 above)
Comment (tranchitella) [reply main.go]: right, done

File: main_test.go (hunk 11 above)
Comment (lluiscampos) [reply main_test.go:19]: Exactly, but we cannot have an *unprotected* SSH key on CI variables. I saw we have it right now and I assumed you added it while developing, but we definitely have to remove before merge, so the tests won't be able to run.

Comment (tranchitella) [reply main_test.go:19]: ok, which options do I have, then? maybe we can discuss it on slack.

Comment (tranchitella) [reply main_test.go:19]: reworked to work with https links when running unit tests

**Comment by tranchitella** [comment]
@lluiscampos I removed the SSH private key variable in GitLab and implemented here (https://github.com/mendersoftware/integration-test-runner/pull/76/commits/be199cb754733167d53125f30157375411f9e34c) the possibility to use HTTPS urls in unit tests.
Can you please review the PR again?

File: util_test.go [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+	"gotest.tools/assert"
+)
+
+func TestGetRemoteURLGitHub(t *testing.T) {
+	url := getRemoteURLGitHub(GitProtocolSSH, "mendersoftware", "workflows")
+	assert.Equal(t, "git@github.com:/mendersoftware/workflows.git", url)
+
+	url = getRemoteURLGitHub(GitProtocolHTTP, "mendersoftware", "workflows")
+	assert.Equal(t, "https://github.com/mendersoftware/workflows", url)
+}
```
Comment (lluiscampos) [inline util_test.go:15]: :+1:

**Review by lluiscampos — APPROVED** [review]
@tranchitella Excellent work. Let's deploy it!
